src/main/java/ch/njol/skript/expressions/ExprTransform.java

```diff
@@ -0,0 +1,168 @@
+	"# {_a::*} is now 1, 2, 3, 4, 5, and 6",
+	"",
+	"# get a list of the sizes of all clans without manually looping",
+	"set {_clan-sizes::*} to indices of {clans::*} transformed using [{clans::%input%::size}]",
+})
+@Since("INSERT VERSION")
+@Keywords("input")
+public class ExprTransform extends SimpleExpression<Object> implements InputSource {
+
+	static {
+		Skript.registerExpression(ExprTransform.class, Object.class, ExpressionType.PATTERN_MATCHES_EVERYTHING,
+				"%~objects% (transformed|mapped) (using|with) \\[<.+>\\]",
+				"%~objects% (transformed|mapped) (using|with) \\(<.+>\\)"
+			);
+		if (!ParserInstance.isRegistered(InputData.class))
+			ParserInstance.registerData(InputData.class, InputData::new);
+	}
+
+	private @UnknownNullability Expression<?> mappingExpr;
+	private @Nullable ClassInfo<?> returnClassInfo;
+	private @UnknownNullability Expression<?> unmappedObjects;
+
+	private final Set<ExprInput<?>> dependentInputs = new HashSet<>();
+
+	private @Nullable Object currentValue;
+	private @UnknownNullability String currentIndex;
+
+	@Override
+	public boolean init(Expression<?>[] expressions, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		unmappedObjects = LiteralUtils.defendExpression(expressions[0]);
+		if (unmappedObjects.isSingle() || !LiteralUtils.canInitSafely(unmappedObjects))
+			return false;
+
+		//noinspection DuplicatedCode
+		if (!parseResult.regexes.isEmpty()) {
+			@Nullable String unparsedExpression = parseResult.regexes.get(0).group();
+			assert unparsedExpression != null;
+			mappingExpr = parseExpression(unparsedExpression, getParser(), SkriptParser.ALL_FLAGS);
+			return mappingExpr != null;
+		}
+		returnClassInfo = Classes.getExactClassInfo(mappingExpr.getReturnType());
+		return true;
+	}
+
+	@Override
+	public @NotNull Iterator<?> iterator(Event event) {
+		if (unmappedObjects instanceof Variable<?>) {
+			Iterator<Pair<String, Object>> variableIterator = ((Variable<?>) unmappedObjects).variablesIterator(event);
+			return StreamSupport.stream(Spliterators.spliteratorUnknownSize(variableIterator, Spliterator.ORDERED), false)
+				.flatMap(pair -> {
+					currentValue = pair.getValue();
+					currentIndex = pair.getKey();
+					return mappingExpr.stream(event);
+				})
+				.iterator();
+		}
+
+		// clear current index just to be safe
+		currentIndex = null;
+
+		Iterator<?> unfilteredObjectIterator = unmappedObjects.iterator(event);
+		if (unfilteredObjectIterator == null)
+			return Collections.emptyIterator();
+		return StreamSupport.stream(Spliterators.spliteratorUnknownSize(unfilteredObjectIterator, Spliterator.ORDERED), false)
+			.flatMap(value -> mappingExpr.stream(event))
+			.iterator();
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event) {
+		try {
+			return Converters.convertStrictly(Iterators.toArray(iterator(event), Object.class), getReturnType());
+		} catch (ClassCastException e1) {
+			return null;
```

Comment:
It might be worth leaving a comment to indicate that you are returning null (rather than an empty array) so that Skript creates one of the right type. You could also just replace this with creating an array of getReturnType.